rewrite-core/src/main/java/org/openrewrite/rpc/RewriteRpc.java

```diff
@@ -0,0 +1,218 @@
+                getObject(sourceFile.getId().toString()) :
+                sourceFile;
+    }
+
+    public <P> VisitResponse scan(SourceFile sourceFile, String visitorName, P p) {
+        return scan(sourceFile, visitorName, p, null);
+    }
+
+    public <P> VisitResponse scan(SourceFile sourceFile, String visitorName, P p,
+                                  @Nullable Cursor cursor) {
+        // Set the local state of this tree, so that when the remote
+        // asks for it, we know what to send.
+        localObjects.put(sourceFile.getId().toString(), sourceFile);
+        String pId = Integer.toString(System.identityHashCode(p));
+
+        Object p2 = p;
+        while (p2 instanceof DelegatingExecutionContext) {
+            p2 = ((DelegatingExecutionContext) p2).getDelegate();
+        }
+        localObjects.put(pId, p2);
+
+        List<String> cursorIds = getCursorIds(cursor);
+
+        return send("Visit", new Visit(visitorName, null, sourceFile.getId().toString(), pId, cursorIds),
+                VisitResponse.class);
+    }
+
+    public Collection<? extends SourceFile> generate(String remoteRecipeId) {
+        List<String> generated = send("Generate", new Generate(remoteRecipeId), GenerateResponse.class);
+        if (!generated.isEmpty()) {
+            return generated.stream()
+                    .map(this::<SourceFile>getObject)
+                    .collect(Collectors.toList());
+        }
+        return emptyList();
+    }
+
+    public List<RecipeDescriptor> getRecipes() {
+        return send("GetRecipes", new GetRecipes(), GetRecipesResponse.class);
+    }
+
+    public Recipe prepareRecipe(String id, Map<String, Object> options) {
+        PrepareRecipeResponse r = send("PrepareRecipe", new PrepareRecipe(id, options), PrepareRecipeResponse.class);
+        return new RpcRecipe(this, r.getId(), r.getDescriptor(), r.getEditVisitor(), r.getScanVisitor());
+    }
+
+    public String print(SourceFile tree) {
+        return print(tree, new Cursor(null, Cursor.ROOT_VALUE));
+    }
+
+    public String print(Tree tree, Cursor parent) {
+        localObjects.put(tree.getId().toString(), tree);
+        return send("Print", new Print(tree.getId().toString(), getCursorIds(parent)), String.class);
+    }
+
+    private @Nullable List<String> getCursorIds(@Nullable Cursor cursor) {
+        List<String> cursorIds = null;
+        if (cursor != null) {
+            cursorIds = cursor.getPathAsStream().map(c -> {
+                String id = c instanceof Tree ?
+                        ((Tree) c).getId().toString()
+                        : Integer.toString(System.identityHashCode(c));
+                localObjects.put(id, c);
+                return id;
+            }).collect(Collectors.toList());
+        }
+        return cursorIds;
+    }
+
+    @VisibleForTesting
+    public <T> T getObject(String id) {
+        RpcReceiveQueue q = new RpcReceiveQueue(remoteRefs, () -> send("GetObject",
```

Comment:
Should we here check if `remoteObjects` already contains the object we are looking for? Not sure about this, but it seems like if we had a map of remote objects that were specific to the current "transaction", we shouldn't have to fetch the object.

Reply:
No, even if the object is in `remoteObjects` it may not be the most recent iteration on that object. We always have to fetch to tell.